Improve error handling in area selection flow

denarch22 opened the pull request:
## 🛠️ Changes
Adds more robust error handling and user feedback for missing or invalid area data in the Search page. Selector now distinguishes between loading, general errors, and empty area types, displaying appropriate messages. This improves the user experience when area data is unavailable or fails to load.

## 📝 Associated issues
Resolves [LIB-206](https://linear.app/biodev/issue/LIB-206/mensaje-cargando-pegado-cuando-no-hay-areas-de-consulta)

denarch22 committed:
Improve error handling in area selection flow

Adds more robust error handling and user feedback for missing or invalid area data in the Search page. Selector now distinguishes between loading, general errors, and empty area types, displaying appropriate messages. This improves the user experience when area data is unavailable or fails to load.

denarch22 committed:
Update README and clean up Selector components

Added instructions for correcting style rules with Prettier in README.md. Refactored Selector.tsx for improved readability and fixed indentation. Removed unnecessary trailing newline in selectorMessages.tsx.

dothedada reviewed (dismissed):
No veo inconveniente en los ajustes que se hacen en el request... por mi lado ok

ManuelStardust commented on src/pages/search/Selector.tsx:
```diff
@@ -54,9 +56,11 @@ const Selector: React.FC<Props> = ({ setShowDrawControl }) => {
+        ? () => <LoadingMessage />
         : areaTypes.length < 1
-        ? LoadingMessage
+        ? () => <ErrorMessage empty={true} />
```
No se si sea bueno pasar este parametro para determinar el mensaje o si es mejor pasar el mensaje directamente al llamar el componente. En tal caso el nombre el parametro deberia llamarse diferente para que indique que se esta evaluando si la lista de areas esta vacia.

erikasv replied:
De acuerdo, me parece mejor pasar el mensaje de error. Porque con la implementación actual solo se pueden manejar dos tipos de errores... y pues ya vimos que tal vez se puedan llegar a necesitar más.

erikasv commented on src/pages/search/Selector.tsx:
```diff
@@ -54,9 +56,11 @@ const Selector: React.FC<Props> = ({ setShowDrawControl }) => {
       component: areasError
-        ? ErrorMessage
+        ? () => <ErrorMessage />
+        : isLoadingAreaTypes
+        ? () => <LoadingMessage />
         : areaTypes.length < 1
-        ? LoadingMessage
+        ? () => <ErrorMessage empty={true} />
```
Por otro lado, tanto operador ternario anidado ya me parece muy confuso.

Se me ocurre manejar el tipo del error en `areasError`, que ya no sea un booleano, podríamos manejar un objeto (para simular un enum). El cual se puede asignar también cuando la promesa se resuelve y además el resultado es un arreglo vacío.

Pero sí me gustaría mantener el estado `isLoadingAreaTypes` y controlar lo que se muestra a partir de los estados `isLoadingAreaTypes` y `areasError`

denarch22 committed:
Improve area error handling in Selector component

Refactored area error state to use a discriminated union for more granular error handling. Updated Selector to display specific error messages for empty results and request failures, and refactored ErrorMessage to accept a message prop for flexibility.

denarch22 committed:
Merge branch 'develop' into darmenta/LIB-206

dothedada commented on src/pages/search/Selector.tsx:
```diff
@@ -18,11 +18,20 @@ interface Props {
-  const [areasError, setAreasError] = useState(false);
+  const [areasError, setAreasError] = useState<AreasErrorType>("none");
   const [polygonError, setPolygonError] = useState(false);
+  const [isLoadingAreaTypes, setIsLoadingAreaTypes] = useState(false);
```
Creo que acá debe ser `true` para el montaje del componente, o, debería setearse como `true` al inico del useEffect para denotar el inicio de carga... creo que no se setea como `true` en ninguna parte y se perdería el feedback de carga para el usuario

erikasv replied:
De acuerdo, actualmente no se estaría mostrando el mensaje de cargando

erikasv replied:
@denarch22 volviste a solicitar revisión, pero este no se resolvió

denarch22 replied:
Si lo realice solo que coloque el set dentro de useEffect

erikasv replied:
perdón, no leí lo del useEffect, la cosa es ese llamado es innecesario, lo mejor es inicializarlo en `true`

dothedada reviewed (commented):
me gustó lo de sacar los mensajes de error y que se desacople el componente de error de los mensajes... solo hay una cosita que vi en el useEffect que en el comentario abordo mejor

erikasv commented on src/pages/search/Selector.tsx:
```diff
@@ -18,11 +18,20 @@ interface Props {
+  const AREA_ERROR_MESSAGES: Record<AreasErrorType, string> = {
+    none: "",
+    "request-failed":
+      "Hubo un error en esta funcionalidad, prueba otra alternativa.",
```
Arreglar indentación

denarch22 replied:
Intente realizar la indentacion para dejarlo en una sola linea, pero prettier me realiza el salto de linea

erikasv replied:
perdón, mi error, leí mal y lo ví mal indentado :sweat_smile:

denarch22 committed:
Fix indentation and loading state in Selector

Corrected indentation in AREA_ERROR_MESSAGES and ensured setIsLoadingAreaTypes(true) is called at the start of useEffect to properly set loading state before fetching area types.

denarch22 committed:
Fix formatting in AREA_ERROR_MESSAGES and whitespace

Corrected formatting of the 'request-failed' error message in AREA_ERROR_MESSAGES and removed unnecessary whitespace in Selector.tsx for improved code readability.

denarch22 committed:
Format AREA_ERROR_MESSAGES for readability

Improved the formatting of the AREA_ERROR_MESSAGES object in Selector.tsx for better readability and maintainability.

denarch22 committed:
Set initial area types loading state to true

The isLoadingAreaTypes state now defaults to true instead of false. This ensures the loading indicator is shown until area types are fetched, improving user feedback during initial load.
